py3status/py3.py

```diff
@@ -1008,3 +1009,165 @@ def request(self, url, params=None, data=None, headers=None,
+            idx -= 1
+
+        return idx
+
+    def _index_to_value(self, index, length, minimum, maximum):
+        """
+        Find the average value of the segment at index
+
+        :param index: index of segment to find the average of
+        :param length: number of segments
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        """
+
+        segment = (maximum - minimum) / length
+        return segment * (index + 0.5)
+
+    def _value_to_bar(self, value, minimum, maximum, bars):
+        """
+        Create a bar of the appropriate height for each value and adjusts the
+        colors depending on the thresholds.
+        There are 8 possible levels.
+        By default, this assumes values are percentages, but the min/max can be
+        adjusted.
+
+        :param value: value to transform to a bar
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        :param bars: characters to use for the different levels of the bar
+
+        :returns: Composite
+        """
+
+        idx = self._value_to_index(value, len(bars), minimum, maximum)
+        full_text = bars[idx]
+        color = self.threshold_get_color(value)
+        return {'full_text': full_text, 'color': color}
+
+    def concurrent_bar_graphs(self, values, minimum=0, maximum=100, bars=None,
+                              separator=''):
+        """
+        Transforms a list of value into a list of bar Composites.
+
+        :param values: list of values to transform into a bar graph
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        :param bars: characters to use for the different levels of the bar
+        :param separator: character to insert between the bars
+
+        :returns: Composite list
+        """
+
+        if bars is None:
+            bars = self._default_bars
+
+        value_bars = []
+        for val in values:
+            value_bars.append(self._value_to_bar(val, minimum, maximum, bars))
+
+        return self.composite_join(separator, value_bars)
+
+    def history_bar_graph(self, history, value, length=1, default=0,
+                          minimum=0, maximum=100, bars=None, separator=''):
+        """
+        Transform a value into a bar graph with history.
+
+        :param history: history of the variable
+        :param value: value to add to the history
+        :param length: length of the history
+        :param default_value: default value to pad with
+        :param minimum: minimum value possible
+        :param maximum: maximum value possible
+        :param bars: characters to use for the different levels of the bar
+        :param separator: character to insert between the bars
+
+        :returns: Composite list
+        """
+
+        if bars is None:
+            bars = self._default_bars
+
+        while len(history) < length:
+            history.append(default)
```

Comment:
`history = [default for i in range(length - 1)]` works?

Reply:
hum. This one is actually tricky. We want to pad the beginning of the history, not the end, but yes, we should just compute how many we need and add them, not loop. That's sloppy.